File: smart_crane/algorithms/post_processing/bezier.py
import logging
import math
from typing import List, Optional, TYPE_CHECKING

from .base import PathPostProcessor, CollisionChecker, NodeType
from smart_crane.core.rust_bridge import RustBackend
from smart_crane.core.constants import (
    SHAPE_CIRCLE,
    DEFAULT_BEZIER_SMOOTHNESS,
    DEFAULT_BEZIER_SEGMENTS,
    DEFAULT_BEZIER_ATTEMPTS,
    BEZIER_CHECK_STEP_MULTIPLIER,
    BEZIER_MIN_STEPS,
    PATH_MIN_NODES,
    RESOLUTION_HALF_FACTOR,
    FOOTPRINT_HALF_DIVISOR,
)

if TYPE_CHECKING:
    from smart_crane.core.map_manager import WorkshopMapManager
    from smart_crane.core.config import Settings


class BezierSmoothProcessor(PathPostProcessor):
    """贝塞尔平滑处理器 (Bezier Curve Smoother)。

    利用二次贝塞尔曲线 (Quadratic Bezier Curve) 将路径中的折线拐角打磨成平滑的圆弧。
    包含自适应回退机制：如果生成的曲线发生碰撞，自动减小曲率半径重试。

    **Rust 加速支持**:
    如果 Rust 核心可用，将调用 `RustPostProcessor.bezier_smooth` 进行高性能计算。
    """

    def __init__(
        self,
        map_mgr: Optional["WorkshopMapManager"] = None,
        settings: Optional["Settings"] = None,
        smoothness: float = DEFAULT_BEZIER_SMOOTHNESS,
        segments: int = DEFAULT_BEZIER_SEGMENTS,
        logger: Optional[logging.Logger] = None,
    ):
        """初始化平滑器。

        Args:
            map_mgr: 地图管理器。
            settings: 全局配置。
            smoothness: 平滑度因子 (0.0 ~ 0.5)。
            segments: 曲线插值段数。
            logger: 父级日志记录器。
        """
        super().__init__(
            name="BezierSmoother", map_mgr=map_mgr, settings=settings, logger=logger
        )

        self.default_smoothness = max(0.0, min(0.5, smoothness))
        self.num_segments = segments

        self.logger.info(
            f"初始化完成. 平滑度={self.default_smoothness}, 插值段数={self.num_segments}"
        )

    def _process_core(
        self, path: List[NodeType], is_safe_fn: CollisionChecker
    ) -> List[NodeType]:
        """执行贝塞尔平滑逻辑。"""
        if not path or len(path) < PATH_MIN_NODES:
            return path

        # =====================================================================
        # 1. 尝试使用 Rust 核心
        # =====================================================================
        if (
            self.map_mgr
            and self.map_mgr.rust_map
            and self.settings
            and RustBackend.is_available()
        ):
            RustProcessor = RustBackend.get_post_processor_class()
            if RustProcessor:
                try:
                    self.stats["mode"] = "rust"
                    # 计算参数 (复用 Greedy 的计算逻辑)
                    shape = self.settings.crane.footprint_shape
                    w = self.settings.crane.footprint_width
                    l = self.settings.crane.footprint_length
                    radius_m = (
                        (w / FOOTPRINT_HALF_DIVISOR)
                        if shape == SHAPE_CIRCLE
                        else (math.hypot(w, l) / FOOTPRINT_HALF_DIVISOR)
                    )
                    radius_m += self.map_mgr.resolution_m * RESOLUTION_HALF_FACTOR
                    z_margin = (
                        self.settings.crane.z_safety_margin
                        + self.settings.crane.footprint_height / FOOTPRINT_HALF_DIVISOR
                    )
                    is_infinite = self.settings.crane.obstacle_infinite_height

                    # 调用 Rust
                    smoothed = RustProcessor.bezier_smooth(
                        path,  # type: ignore
                        self.map_mgr.rust_map,
                        self.default_smoothness,
                        self.num_segments,
                        radius_m,
                        z_margin,
                        is_infinite,
                        path[0],  # grace_start
                        path[-1],  # grace_end
                    )
                    return smoothed

                except Exception as e:
                    self.logger.error(f"Rust 贝塞尔平滑失败，降级至 Python: {e}")
                    # Fallback...

        # =====================================================================
        # 2. Python 原生实现 (Fallback)
        # =====================================================================
        self.stats["mode"] = "python"
        smoothed_path = []
        smoothed_path.append(path[0])

        input_nodes_count = len(path)
        curved_corners = 0

        for i in range(1, len(path) - 1):
            p0 = path[i - 1]
            p1 = path[i]
            p2 = path[i + 1]

            current_smoothness = self.default_smoothness
            best_segment = None

            for attempt in range(DEFAULT_BEZIER_ATTEMPTS):
                if self._check_curve_safety(p0, p1, p2, current_smoothness, is_safe_fn):
                    best_segment = self._generate_curve(p0, p1, p2, current_smoothness)
                    if attempt > 0:
                        self.logger.debug(
                            f"拐角 {i} 触发自适应回退: 尝试 {attempt} 次后成功 "
                            f"(Smoothness: {self.default_smoothness} -> {current_smoothness:.4f})"
                        )
                    break
                else:
                    current_smoothness *= 0.5

            if best_segment:
                smoothed_path.extend(best_segment)
                curved_corners += 1
            else:
                self.logger.debug(f"拐角 {i} 空间狭窄无法平滑，保留硬连接。")
                smoothed_path.append(p1)

        smoothed_path.append(path[-1])

        self.logger.debug(
            f"平滑统计: 处理 {curved_corners}/{input_nodes_count-2} 个拐角"
        )
        return smoothed_path

    def _check_curve_safety(
        self,
        p0: NodeType,
        p1: NodeType,
        p2: NodeType,
        s: float,
        is_safe_fn: CollisionChecker,
    ) -> bool:
        """检查生成的贝塞尔曲线是否安全。"""
        dims = len(p0)
        q0 = tuple(p1[d] + (p0[d] - p1[d]) * s for d in range(dims))
        q2 = tuple(p1[d] + (p2[d] - p1[d]) * s for d in range(dims))

        d1 = math.sqrt(sum((p1[d] - p0[d]) ** 2 for d in range(dims)))
        d2 = math.sqrt(sum((p2[d] - p1[d]) ** 2 for d in range(dims)))
        total_len = (d1 + d2) * s

        steps = int(total_len * BEZIER_CHECK_STEP_MULTIPLIER)
        steps = max(steps, BEZIER_MIN_STEPS)

        for k in range(steps + 1):
            t = k / steps
            check_point_list = []
            for d in range(dims):
                val = (1 - t) ** 2 * q0[d] + 2 * t * (1 - t) * p1[d] + t**2 * q2[d]
                check_point_list.append(val)
            check_point = tuple(check_point_list)
            if not is_safe_fn(check_point):  # type: ignore
                return False
        return True

    def _generate_curve(
        self, p0: NodeType, p1: NodeType, p2: NodeType, s: float
    ) -> List[tuple]:
        """生成贝塞尔曲线点集。"""
        dims = len(p0)
        q0 = tuple(p1[d] + (p0[d] - p1[d]) * s for d in range(dims))
        q2 = tuple(p1[d] + (p2[d] - p1[d]) * s for d in range(dims))

        points = []
        # 必须从 i=0（即 t=0 的 q0）开始输出。
        # _check_curve_safety 是从 t=0 起做碰撞检查的，若这里跳过 q0，
        # 实际路径就变成"上一点 → curve(1/n)"这条弦——它比校验过的
        # "q0 → curve(1/n)"更贴近拐角内侧，且从未被任何检查覆盖。
        # q0 位于 p0→p1 线段上，与上一段的终点不重合，不会产生重复点。
        for i in range(0, self.num_segments + 1):
            t = i / self.num_segments
            pt_list = []
            for d in range(dims):
                val = (1 - t) ** 2 * q0[d] + 2 * t * (1 - t) * p1[d] + t**2 * q2[d]
                pt_list.append(val)
            points.append(tuple(pt_list))
        return points

File: src/post_processing/bezier.rs
use crate::map::RustMapManager;
use log::debug;

/// 贝塞尔平滑处理器实现 (Bezier Curve Smoother Implementation)。
///
/// 逻辑与 Python 版 `BezierSmoothProcessor` 保持一致。
pub struct BezierSmoother;

impl BezierSmoother {
    /// 执行贝塞尔平滑。
    ///
    /// # Arguments
    /// * `path` - 原始路径点列表。
    /// * `map` - 地图管理器引用。
    /// * `smoothness` - 初始平滑度 (0.0 - 0.5)。
    /// * `segments` - 曲线插值段数。
    /// * `check_radius` - 碰撞检测半径。
    /// * `check_z_margin` - 垂直安全边距。
    /// * `ignore_z` - 是否忽略高度检测。
    /// * `grace_start` - 豁免区起点。
    /// * `grace_end` - 豁免区终点。
    // 参数均为彼此独立的物理量（尺寸、分辨率、边距、阈值），
    // 打包成 struct 只会增加一层间接，不提升可读性。
    #[allow(clippy::too_many_arguments)]
    pub fn process(
        path: Vec<(f32, f32, f32)>,
        map: &RustMapManager,
        smoothness: f32,
        segments: usize,
        check_radius: f32,
        check_z_margin: f32,
        ignore_z: bool,
        grace_start: (f32, f32, f32),
        grace_end: (f32, f32, f32),
    ) -> Vec<(f32, f32, f32)> {
        if path.len() < 3 {
            return path;
        }

        let default_smoothness = smoothness.clamp(0.0, 0.5);
        let mut smoothed_path = Vec::new();
        smoothed_path.push(path[0]);

        let input_nodes_count = path.len();
        let mut curved_corners = 0;

        // 遍历每一个拐角点 P1 (从第2个点到倒数第2个点)
        for i in 1..path.len() - 1 {
            let p0 = path[i - 1];
            let p1 = path[i];
            let p2 = path[i + 1];

            let mut current_smoothness = default_smoothness;
            let mut best_segment: Option<Vec<(f32, f32, f32)>> = None;

            // 自适应回退机制 (Adaptive Fallback)
            // 尝试生成曲线，如果碰撞则减半平滑度重试，最多 5 次
            for attempt in 0..5 {
                if Self::check_curve_safety(
                    p0,
                    p1,
                    p2,
                    current_smoothness,
                    map,
                    check_radius,
                    check_z_margin,
                    ignore_z,
                    grace_start,
                    grace_end,
                ) {
                    best_segment = Some(Self::generate_curve(
                        p0,
                        p1,
                        p2,
                        current_smoothness,
                        segments,
                    ));

                    if attempt > 0 {
                        debug!(
                            "[Rust Bezier] 拐角 {} 触发自适应回退: 尝试 {} 次后成功 (S: {:.4} -> {:.4})",
                            i,
                            attempt,
                            default_smoothness,
                            current_smoothness
                        );
                    }
                    break;
                } else {
                    current_smoothness *= 0.5;
                }
            }

            if let Some(segment) = best_segment {
                smoothed_path.extend(segment);
                curved_corners += 1;
            } else {
                debug!("[Rust Bezier] 拐角 {} 空间狭窄无法平滑，保留硬连接。", i);
                smoothed_path.push(p1);
            }
        }

        smoothed_path.push(path[path.len() - 1]);

        debug!(
            "[Rust Bezier] 平滑统计: 处理 {}/{} 个拐角",
            curved_corners,
            input_nodes_count - 2
        );

        smoothed_path
    }

    /// 检查生成的贝塞尔曲线是否安全。
    #[allow(clippy::too_many_arguments)]
    fn check_curve_safety(
        p0: (f32, f32, f32),
        p1: (f32, f32, f32),
        p2: (f32, f32, f32),
        s: f32,
        map: &RustMapManager,
        radius: f32,
        z_margin: f32,
        ignore_z: bool,
        grace_start: (f32, f32, f32),
        grace_end: (f32, f32, f32),
    ) -> bool {
        // 计算切点 (Anchor Points)
        let q0 = (
            p1.0 + (p0.0 - p1.0) * s,
            p1.1 + (p0.1 - p1.1) * s,
            p1.2 + (p0.2 - p1.2) * s,
        );
        let q2 = (
            p1.0 + (p2.0 - p1.0) * s,
            p1.1 + (p2.1 - p1.1) * s,
            p1.2 + (p2.2 - p1.2) * s,
        );

        // 估算曲线弧长
        let d1 = ((p1.0 - p0.0).powi(2) + (p1.1 - p0.1).powi(2) + (p1.2 - p0.2).powi(2)).sqrt();
        let d2 = ((p2.0 - p1.0).powi(2) + (p2.1 - p1.1).powi(2) + (p2.2 - p1.2).powi(2)).sqrt();
        let total_len = (d1 + d2) * s;

        // 采样密度: 每 0.2m 检测一次
        let steps = (total_len * 5.0) as i32;
        let steps = steps.max(5);

        for k in 0..=steps {
            let t = k as f32 / steps as f32;
            let t_sq = t * t;
            let inv_t = 1.0 - t;
            let inv_t_sq = inv_t * inv_t;

            // 二阶贝塞尔公式: B(t) = (1-t)^2 * q0 + 2t(1-t) * p1 + t^2 * q2
            let cx = inv_t_sq * q0.0 + 2.0 * t * inv_t * p1.0 + t_sq * q2.0;
            let cy = inv_t_sq * q0.1 + 2.0 * t * inv_t * p1.1 + t_sq * q2.1;
            let cz = inv_t_sq * q0.2 + 2.0 * t * inv_t * p1.2 + t_sq * q2.2;

            // 豁免区检测
            let d_start_sq = (cx - grace_start.0).powi(2)
                + (cy - grace_start.1).powi(2)
                + (cz - grace_start.2).powi(2);
            if d_start_sq < 0.25 {
                continue;
            }
            let d_end_sq = (cx - grace_end.0).powi(2)
                + (cy - grace_end.1).powi(2)
                + (cz - grace_end.2).powi(2);
            if d_end_sq < 0.25 {
                continue;
            }

            if map.check_collision_raw(cx, cy, cz, radius, z_margin, ignore_z) {
                return false;
            }
        }
        true
    }

    /// 生成贝塞尔曲线点集。
    fn generate_curve(
        p0: (f32, f32, f32),
        p1: (f32, f32, f32),
        p2: (f32, f32, f32),
        s: f32,
        segments: usize,
    ) -> Vec<(f32, f32, f32)> {
        let q0 = (
            p1.0 + (p0.0 - p1.0) * s,
            p1.1 + (p0.1 - p1.1) * s,
            p1.2 + (p0.2 - p1.2) * s,
        );
        let q2 = (
            p1.0 + (p2.0 - p1.0) * s,
            p1.1 + (p2.1 - p1.1) * s,
            p1.2 + (p2.2 - p1.2) * s,
        );

        let mut points = Vec::with_capacity(segments + 1);

        // 从 i=0（t=0，即 q0）开始输出，包含终点 q2。
        //
        // 早期实现从 1..=segments 起，理由是"q0 是上一段的终点，避免重复"——
        // 但这不成立：q0 位于 P0→P1 线段上、距 P1 为 s 的位置，与上一段的
        // 终点（上一拐角的 q2）并不重合。跳过它会让实际路径变成
        // "上一点 → curve(1/n)"这条弦，它比碰撞检查覆盖的
        // "q0 → curve(1/n)"更贴近拐角内侧，且从未被验证。
        // 碰撞检查是从 t=0 起做的，输出必须与之一致。
        for i in 0..=segments {
            let t = i as f32 / segments as f32;
            let t_sq = t * t;
            let inv_t = 1.0 - t;
            let inv_t_sq = inv_t * inv_t;

            let px = inv_t_sq * q0.0 + 2.0 * t * inv_t * p1.0 + t_sq * q2.0;
            let py = inv_t_sq * q0.1 + 2.0 * t * inv_t * p1.1 + t_sq * q2.1;
            let pz = inv_t_sq * q0.2 + 2.0 * t * inv_t * p1.2 + t_sq * q2.2;

            points.push((px, py, pz));
        }

        points
    }
}
